# Pandas study mini project

## Tasks

1. Import the `pandas` library as `pd`. Upload [bookings.csv](files/bookings.csv) dataset 
        with separator - `;` .\
        Check the table size, variable types, and then print the first 7 lines 
        to look at the data.
2. Change the column names to lowercase and replace spaces with underscores. 
3. Users from which countries have made the most successful bookings? 
        Specify the top 5.
4. How many nights on average are booked in different types of hotels
5. Sometimes the type of room received by the client ( `assigned_room_type` )\
        differs from the one originally booked ( `reserved_room_type` ).\
        This can happen, for example, due to overbooking. 
        How many such observations were found in the dataset?
6. Review your planned arrival dates.\
        - What month was the most successful booking in 2016?
            Has the most popular month changed in 2017?\
        - Group the data by years and check which month City Hotel bookings 
            were canceled the most in each period.
7. Look at the numerical characteristics of the three variables: 
        `adults`, `children` and `babies`.\
        Which one has the largest average value?
8. Create a `total_kids` column by concatenating `children` and `babies`.\
        What types of hotels, on average, are more popular with clients with children?
9. Create a `has_kids` variable that takes the value `True`\
        if the client specified 
        at least one child during the booking ( `total_kids`),\
        and `False` otherwise.\
        Calculate the ratio of the number of users who left to 
        the total number of customers,\
        expressed as a percentage (`churn rate`). 
        Indicate which group has the highest rate.

## Data Description

**There are the following variables:**

* Hotel - hotel type (City Hotel or Resort Hotel)
* Is canceled - the booking was canceled (1) or not (0); 
        what is not canceled is considered successful
* Lead time - the number of days elapsed between the date of booking 
        and the date of arrival 
* Arrival full date
* Arrival date year - year of arrival
* Arrival date month - month of arrival
* Arrival date week number - arrival week number
* Arrival date day of month - day of arrival
* Stays in weekend nights - the number of days off (Saturday or Sunday) 
        that the guest has booked to stay at the hotel
* Stays in week nights - the number of days (from Monday to Friday) 
        that the guest has booked to stay at the hotel
* Stays total nights - total number of nights booked 
        (sum of the two previous columns)
* Adults - number of adults
* Children - number of children
* Babies - number of babies
* Meal - selected type of food
* Country - client's country of origin
* Reserved room type - reserved room type
* Assigned room type - type of room received (may differ from booked)
* Customer type - booking type
* Reservation status - value of the last reservation status: 
        Canceled - was canceled by the client; 
        Check-Out - the client has checked in, but has already left the hotel; 
        No-Show - the client did not register and informed 
        the hotel administration the reason
* Reservation status date - Status update date

In [50]:
import pandas as pd

In [51]:
df = pd.read_csv('./bookings.csv', sep=';')

In [57]:
df.shape

(119390, 21)

In [58]:
df.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

In [60]:
df.head(7)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

In [63]:
df.columns

Index(['Hotel', 'Is Canceled', 'Lead Time', 'arrival full date',
       'Arrival Date Year', 'Arrival Date Month', 'Arrival Date Week Number',
       'Arrival Date Day of Month', 'Stays in Weekend nights',
       'Stays in week nights', 'stays total nights', 'Adults', 'Children',
       'Babies', 'Meal', 'Country', 'Reserved Room Type', 'Assigned room type',
       'customer type', 'Reservation Status', 'Reservation status_date'],
      dtype='object')

In [73]:
# lower-text to lowercase, strip-remove spaces, replace-replace
df.columns = df.columns.str.lower() \
                       .str.strip() \
                       .str.replace(' ', '_')
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_full_date',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'stays_total_nights', 'adults', 'children',
       'babies', 'meal', 'country', 'reserved_room_type', 'assigned_room_type',
       'customer_type', 'reservation_status', 'reservation_status_date'],
      dtype='object')

In [96]:
# Users from which countries have made the most successful bookings? Specify the top 5.
df.query('is_canceled == 0').country.value_counts()[:5]

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country, dtype: int64

In [114]:
# How many nights on average are booked in different types of hotels
df.groupby('hotel', as_index=False) \
    .agg({'stays_total_nights': 'mean'})\
    .round(2)

hotel  stays_total_nights
0    City Hotel                2.98
1  Resort Hotel                4.32

**Average number of nights booked at City Hotel was 2.98. Average number of nights booked at Resort Hotel was 4.32**

In [118]:
# assigned_room_type != reserved_room_type ;
# How many such observations were found in the dataset?
df.query('assigned_room_type != reserved_room_type').shape[0]

14917

**14917 observations where reserved and assigned room types differed**

In [123]:
# Review your planned arrival dates.
# - What month was the most successful booking in 2016? Has the most popular month changed in 2017?
# - Group the data by years and check which month City Hotel bookings were canceled the most in each period.

df.head()
df.query('arrival_date_year == 2016').arrival_date_month.value_counts()[:1]

October    6203
Name: arrival_date_month, dtype: int64

In [124]:
df.query('arrival_date_year == 2017').arrival_date_month.value_counts()[:1]

May    6313
Name: arrival_date_month, dtype: int64

In [134]:
df \
    .query('hotel == "City Hotel" & is_canceled == 1') \
    .groupby(['arrival_date_year', 'arrival_date_month'], as_index=False) \
    .agg({'is_canceled' : 'count'}) \
    .sort_values('is_canceled', ascending=False)

arrival_date_year arrival_date_month  is_canceled
25               2017                May         2217
16               2016            October         1947
18               2017              April         1926
23               2017               June         1808
12               2016               June         1720
17               2016          September         1567
5                2015          September         1543
6                2016              April         1539
14               2016                May         1436
15               2016           November         1360
22               2017               July         1324
4                2015            October         1321
24               2017              March         1278
7                2016             August         1247
0                2015             August         1232
19               2017             August         1123
13               2016              March         1108
8                2016           December         1072
21               2017            January         1044
11               2016               July         1043
20               2017           February          971
2                2015               July          939
9                2016           February          930
1                2015           December          668
10               2016            January          438
3                2015           November          301

In [137]:
# OR df.groupby('column_1')['column_2'].what_to_do()
df.query('is_canceled == 1 & hotel == "City Hotel"')\
    .groupby('arrival_date_year')['arrival_date_month'].value_counts()

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1

**October was the most successful booking month of 2016. May was the most successful booking month of 2017\
September was the most canceled booking month of 2015. October was the most canceled booking month of 2016.\
May was the most canceled booking month of 2017**

In [140]:
# Look at the numerical characteristics of the three variables: adults, children and babies.
# Which one has the largest average value?
df.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights  stays_total_nights  \
count            119390.000000         119390.000000       119390.000000   
mean                  0.927599              2.500302            3.427900   
std                   0.998613              1.908286            2.557439   
min                   0.000000              0.000000            0.000000   
25%                   0.000000              1.000000            2.000000   
50%                   1.000000              2.000000            3.000000   
75%                   2.000000              3.000000            4.000000   
max                  19.000000             50.000000           69.000000   

              adults       children         babies  
count  119390.000000  119386.000000  119390.000000  
mean        1.856403       0.103890       0.007949  
std         0.579261       0.398561       0.097436  
min         0.000000       0.000000       0.000000  
25%         2.000000       0.000000       0.000000  
50%         2.000000       0.000000       0.000000  
75%         2.000000       0.000000       0.000000  
max        55.000000      10.000000      10.000000

In [147]:
# Create a total_kids column by concatenating children and babies.
# What types of hotels, on average, are more popular with clients with children?
df['total_kids'] = df['children'] + df['children']
df.groupby('hotel', as_index=False) \
    .agg({'total_kids' : 'mean'}).round(2)

hotel  total_kids
0    City Hotel        0.18
1  Resort Hotel        0.26

***Resort hotel type is more popular with clients with children***

In [161]:
# Create a has_kids variable that takes the value True
# if the client specified at least one child during the booking ( total_kids),
# and False otherwise.
# Calculate the ratio of the number of users who left to the total number of customers,
# expressed as a percentage (churn rate). Indicate which group has the highest rate.
df['has_kids'] = df['total_kids'] > 0
no_kids = df.query('has_kids == False and is_canceled == 1').shape[0]/df.query('has_kids == False').shape[0]
no_kids = round(no_kids*100, 2)
no_kids

37.09

In [163]:
yes_kids = df.query('has_kids == True and is_canceled == 1').shape[0]/df.query('has_kids == True').shape[0]
yes_kids = round(yes_kids*100, 2)
yes_kids

36.39

***the churn rate of clients who have children is higher than that of those who do not have children, at 37.09%***